### PR TITLE
Change `lambda` helper to take parameter names from block

### PR DESCRIPTION
**Pull request by andy-faunadb**

### CODE REVIEW

File: lib/fauna/query.rb [hunk 1]
```diff
@@ -457,15 +457,11 @@ def self.not(boolean)
...
-      next_var_number = low_var_number + n_vars
-
-      Thread.current[:fauna_lambda_var_number] = next_var_number
-
-      yield (low_var_number...next_var_number).map { |i| "auto#{i}" }
-    ensure
-      Thread.current[:fauna_lambda_var_number] = low_var_number
+    def self.block_parameters(block)
+      block.parameters.map do |kind, name|
+        fail ArgumentError, 'All lambda parameters must be non-rest.' if kind == :rest
```
Comment (freels) [inline lib/fauna/query.rb]: 'splat' is the right nomenclature: "Splat parameters are not supported in lambda expressions."

File: lib/fauna/query.rb [hunk 2]
```diff
@@ -89,15 +89,15 @@ def self.quote(expr)
...
-          lambda_expr vars[0], block.call(var(vars[0]))
-        else
-          lambda_expr vars, block.call(*(vars.map { |v| var(v) }))
-        end
+      vars = block_parameters block
+      n_args = vars.length
+      case n_args
+      when 0
+        fail ArgumentError, 'Block must take at least 1 argument.'
+      when 1
```
Comment (freels) [inline lib/fauna/query.rb]: Since this case is redundant, I'd just take it out.

Comment (andy-faunadb) [inline lib/fauna/query.rb]: We don't use an array pattern in the case of only one var. `:foo`, not `[:foo]`.

Comment (freels) [inline lib/fauna/query.rb]: oh, good point, I missed the distinction.

**Comment by freels** [comment]
shipit

File: lib/fauna/query.rb [hunk 3]
```diff
@@ -89,15 +89,16 @@ def self.quote(expr)
...
-
-      with_auto_vars(n_args) do |vars|
-        if n_args == 1
-          lambda_expr vars[0], block.call(var(vars[0]))
-        else
-          lambda_expr vars, block.call(*(vars.map { |v| var(v) }))
-        end
+      vars = block_parameters block
+      n_args = vars.length
+      case n_args
```
Comment (ashfire908) [inline lib/fauna/query.rb]: Couldn't we just do `case vars.length` on the next line and drop assigning that to `n_args`?

**Pull request merged**